utils/forest_utils/src/io/writer_checksum.rs

```diff
@@ -3,24 +3,24 @@
 use std::{marker::PhantomData, pin::Pin, task::Poll};
 
 use digest::{Digest, Output};
+use futures::AsyncWrite;
 use pin_project_lite::pin_project;
-use tokio::io::AsyncWrite;
 
 pin_project! {
-    /// Wrapper `AsyncWriter` implementation that calculates the checksum on the fly.
+    /// Wrapper `AsyncWriter` implementation that calculates the optional checksum on the fly.
     /// Both `Writer` and `Digest` parameters are generic so one can use freely the relevant
     /// structures, e.g. `BufWriter` and `Sha256`.
     pub struct AsyncWriterWithChecksum<D, W> {
         #[pin]
         inner: W,
-        hasher: D,
+        hasher: Option<D>,
```

Comment:
I'm not sure about it. Having `AsyncWriterWithChecksum` without checksum makes it just an `AsyncWriter`. This introduces hasher checks on every write. Do you think we can do it in a less invasive way?

Reply:
I've simplified the checksum logic to only use `AsyncWriterWithOptionalChecksum` inside `export` fn, let me know if that's ok or did u suggest anything else

Reply:
OK if u don't like `Option<Hasher>`, maybe I can restore the logic and do `Either<AsyncWriter, AsyncWriterWithChecksum>`, sounds good?

Reply:
I'm fine with both, provided the solution with `Option<Hasher>` doesn't get a performance hit (can be checked by exporting mainnet snapshot).

Reply:
Does this mean an additional `if` condition makes it faster? We should put `if`s everywhere :rofl: 

Anyway, it's fine.